Update the docs for JobManagedBy and JobPodReplacementPolicy related to pod termination

mimowo opened the pull request:
Tracking issue: https://github.com/kubernetes/enhancements/issues/4368.
The change of the logic reflected in the notes describes the fix for the issue: 
 https://github.com/kubernetes/kubernetes/issues/123775

<!--

 Hello!

 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/suggesting-improvements/

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.

-->

dipesh-rawat commented:
/retitle [WIP] Update the docs for JobManagedBy and other features related by the fix

dipesh-rawat commented:
/sig apps

sftim commented on content/en/docs/concepts/workloads/controllers/job.md:
````diff
@@ -1065,6 +1065,11 @@ status:
 ```
 
+{{< note >}}
+Since v1.31 the `JobPodReplacementPolicy` feature gate guards the fix for the issue:
````
We prefer to link to the list of feature gates:
```suggestion
Since v1.31 the `JobPodReplacementPolicy` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) guards the fix for the issue:
```

mimowo replied:
For sure, it was just a quick note. I m going to work on a proper documentation update this week.

mimowo replied:
Updated. PTAL

sftim reviewed (commented):
I don't understand the change being made here. Can we make it easier for people who are not K8s contributors to understand what is being explained?

hacktivist123 commented:
Hello @mimowo 👋! I'm reaching out from the Docs team. Just checking in as we approach Docs Freeze on Tuesday July 30th 18:00 PDT. This documentation appears to still be under review. To meet the Docs Freeze, this PR must have a technical review as well as lgtm and approve labels applied, without any unaddressed comments or concerns from SIG Docs. The status of this enhancement is marked as at risk for docs freeze. Thank you!

hacktivist123 commented:
/sig apps

mimowo commented:
@hacktivist123 thanks for reaching out. I will try to update the PR this week and ask for reviews.

mimowo commented:
/cc @atiratree @soltysh 
For review from the sig-apps POV. 
/cc @sftim 
for the update to your commets

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +659,30 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+and thus adding the terminal the Job condition, even if the fate of the Job is
+already determined.
+
+If you want to know, or the controller you write, the fate of the Job as soon
+as determined you can use, since v1.31, the `FailureTarget` and `SuccessCriteriaMet`
+conditions, which cover all scenarios in which Job controller triggers
+the Job termination process.
+
```
crosslinking a discussion https://github.com/kubernetes/website/pull/46807#discussion_r1691543157

mimowo replied:
+1. I have expanded on the `FailureTarget` and `SuccessCriteriaMet` conditions in the recent update. PTAL.

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +659,30 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+indicating that the Job is finished (either `Complete` or `Failed`).
+
+Note that, the process of terminating all pods may take a substantial amount
+of time, depending the one Pod's `terminationGracePeriodSeconds` (see
```
```suggestion
of time, depending on a Pod's `terminationGracePeriodSeconds` (see
```

mimowo replied:
Done

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +659,30 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+Note that, the process of terminating all pods may take a substantial amount
+of time, depending the one Pod's `terminationGracePeriodSeconds` (see
+[Pod termination](#docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)),
+and thus adding the terminal the Job condition, even if the fate of the Job is
```
```suggestion
and thus adding the terminal Job condition, even if the fate of the Job is
```

mimowo replied:
Done

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +659,30 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+and thus adding the terminal the Job condition, even if the fate of the Job is
+already determined.
+
+If you want to know, or the controller you write, the fate of the Job as soon
```
not sure if we need to specify the controller bit here

mimowo replied:
Yeah, me neither, might be just noise. Dropped it.

mimowo commented:
/cc @tenzen-y 
For another pair of sig-apps eyes on it.

tenzen-y commented on content/en/docs/concepts/workloads/controllers/job.md:
````diff
@@ -444,6 +444,10 @@ kubectl get -o yaml job job-backoff-limit-per-index-example
       type: Failed
 ```
 
+Note that, since v1.31, you will also observe in the status the `FailureTarget`
````
IIIUC, the Hugo shortcode would be better as you can see here: https://kubernetes.io/docs/contribute/style/style-guide/#shortcodes

tengqm replied:
`{{< skew currentVersion >}}`

tenzen-y commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +668,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+Job controller awaits for termination of all pods before adding a condition
+indicating that the Job is finished (either `Complete` or `Failed`).
+
+Note that, the process of terminating all pods may take a substantial amount
```
Additionally, using the Hugo short code would be better, right?
But I'm not sure if we should use the Hugo shortcode in all places where we want to mention the Note.

mimowo replied:
Actually, the paragraph seems pretty important for the section so tentatively I remove the "Note that", and don't add the Hugo either. PTAL.

tenzen-y replied:
SGTM

tenzen-y commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +668,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+or `Failed`, respectively.
+
+The failure scenarios encompass:
+- the `.spec.backoffLimit`
```
```suggestion
- the `.spec.backoffLimit` is reached
```

Maybe a full sentence would be better as similar to "the .spec.completions is reached".

mimowo replied:
Yeah, but I think in case of backoffLimit "exceeded" is better. This is also what the reason on the Failed condition says.

tenzen-y replied:
That sounds more reasonable to me. Thanks!

tenzen-y reviewed (commented):
overall lgtm

tengqm commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +668,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+- the `.spec.backoffLimit`
+- the `.spec.activeDeadlineSeconds` is exceeded
+- the `.spec.backoffLimitPerIndex` is exceeded (see [Backoff limit per index](#backoff-limit-per-index))
+- the Pod matches the Job Pod Failure Policy rule with the `FailJob` action  (see more [Pod failure policy](#pod-failure-policy))
```
The spelling "Job Pod Failure Policy" doesn't conform to the website doc style.

mimowo replied:
Updated to use the field name. PTAL.

tengqm commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +668,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+
+The success scenarios encompass:
+- the `.spec.completions` is reached
+- the criteria specified by the Job Success Policy are met (see more [Success policy](#success-policy))
```
"Job Success Policy" should be "Job success policy"?
"Success policy" should be "success policy"?

tenzen-y replied:
```suggestion
- the criteria specified by the `.spec.successPolicy` are met (see more [Success policy](#success-policy))
```

Maybe, field name would be better. WDYT? @mimowo @tengqm

tengqm replied:
using generic terms or field names are both okay to me.

mimowo replied:
Cool, I use the field name in this case here.

tengqm commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +668,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+still remain running at that point in time.
+
+Since v1.31, when you enable either the `JobManagedBy` or
+`JobPodReplacementPolicy` (enabled by default)
```
```suggestion
the `JobPodReplacementPolicy` (enabled by default)
```

mimowo replied:
Thanks, done

mimowo commented:
Thanks everyone for your reviews! The so-far comments are all addressed in https://github.com/kubernetes/website/pull/46808/commits/0318c3f94a64fec995498ca0a651783fda622c9c. PTAL.

tenzen-y reviewed (commented):
Thanks!
/lgtm

@atiratree Could you take another look as a sig-apps approver?

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +670,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+
+Prior to {{< skew currentVersion >}} the Job terminal conditions are added when the Job termination
+process is triggered, and all Pod finalizers are removed, but some pods may
+still remain running at that point in time.
```
```suggestion
still remain running/terminating at that point in time.
```

?

mimowo replied:
sgtm, updated

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +670,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+
+### Termination of Job pods
+
+Prior to {{< skew currentVersion >}} the Job terminal conditions are added when the Job termination
```
how does the `{{< skew currentVersion >}}` work? What happens when we release 1.32?

mimowo replied:
@tengqm can you help with that?

mimowo replied:
Ok, I tested this by looking at other places of its usage. For example this line "The kubelet in Kubernetes v1.30 collects spans from the garbage collection, pod synchronization routine as well as" in https://kubernetes.io/docs/concepts/cluster-administration/system-traces/ was added a year ago and it says "1.30", so there is no any mechanism to freeze the version.

I think our intention is to say "since v1.31", so probably it is better to hard-code it. Alternatively we can say "In Kubernetes {current version} ...". Which would you prefer @tengqm @atiratree ?

mimowo replied:
Maybe the downside is that the documentation contains references to some very old versions like 1.21. 

OTOH, it is nice to know if some feature was introduced in 1.29 or 1.30, for example, if you are considering an upgrade.

atiratree replied:
I don't mind hard-coding it. Currently it would give a wrong implication in 1.32, that one needs to upgrade to latest to observe the behavior.

I suppose it would be good to remove the version mention in some number of releases as the version becomes less relevant. But that it quite hard to plan.

mimowo replied:
I agree, the wording `Prior to {{< skew currentVersion >}}` or  `Since {{< skew currentVersion >}}` will be misleading. Tentatively I use hard-coding now.

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
````diff
@@ -1063,6 +1113,13 @@ status:
 ```
 
+{{< note >}}
+Since {{< skew currentVersion >}}, when you enable the `JobPodReplacementPolicy`
````
same question about the version in other places as well

mimowo replied:
Yeah, let's track the discussion there: https://github.com/kubernetes/website/pull/46808#discussion_r1692964346, and I will update all places if needed.

mimowo commented:
> * condition update: https://github.com/kubernetes/website/pull/46807/files#r1691615551

Done, added the condition with a comment. I also added a note for this before below the example.

mimowo commented:
> * pod-failure-policy.md update: https://github.com/kubernetes/website/pull/46807/files#r1691543157

Good point, actually for FailJob action the condition would be also added before so I don't make the note conditional on version. I think we didn't mention the condition before because back then it was considered more of an "internal detail". So,  it makes sense to extend the documentation on this condition as we make it more official in this release.

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +670,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+The failure scenarios encompass:
+- the `.spec.backoffLimit` is exceeded
+- the `.spec.activeDeadlineSeconds` is exceeded
+- the `.spec.backoffLimitPerIndex` is exceeded (see [Backoff limit per index](#backoff-limit-per-index))
```
Good point, added and rephrased the descriptions of existing ones. PTAL.

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +670,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+- the `.spec.backoffLimit` is exceeded
+- the `.spec.activeDeadlineSeconds` is exceeded
+- the `.spec.backoffLimitPerIndex` is exceeded (see [Backoff limit per index](#backoff-limit-per-index))
+- the Pod matches a rule in `.spec.podFailurePolicy` with the `FailJob` action (see more [Pod failure policy](#pod-failure-policy))
```
> and mention here that `FailIndex` action can speed up the potential job failure when `backoffLimitPerIndex` is used?

I feel this level of detail should go to the dedicated sections. I prefer to keep here just one-liner descirptions, and more details in dedicated sections

mimowo replied:
> Or just add that pod failure policy can further modify the failure scenarios?

Hm, you mean for example `action: Ignore` can influence handling of `spec.backoffLimit`? Yes, there are such cases, but I would prefer to keep the points here as concise one-liners. The users can find out more nuances when going to the specific sections. Capturing all nuances here will be essentially repeating the sections.

mimowo replied:
I have rephrased the descriptions trying to keep the balance between being too concise and going into too much details. PTAL.

atiratree replied:
We do not have to go too much into details. What about something like https://github.com/kubernetes/website/pull/46808#discussion_r1693183116 ?

mimowo replied:
Great. accepted the suggestion as it sounds good to me

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +670,45 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+- the Pod matches a rule in `.spec.podFailurePolicy` with the `FailJob` action (see more [Pod failure policy](#pod-failure-policy))
+
+The success scenarios encompass:
+- the `.spec.completions` is reached
```
Yeah, I prefer the points here to be concise, but I agree the initial working was too cryptic. I have rephrased the points. PTAL.

atiratree replied:
Thx

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,51 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+- the Job runtime exceeded the specified `.spec.activeDeadlineSeconds`
+- the Job using `.spec.backoffLimitPerIndex` has some indexes failed;
+  see also [Backoff limit per index](#backoff-limit-per-index)
+- the number of failed indexes in the Job using has exceeded the specified
+  `spec.maxFailedIndexes`; see also [Backoff limit per index](#backoff-limit-per-index)
```
```suggestion
- the number of failed indexes in the Job has exceeded the specified
  `spec.maxFailedIndexes`; see also [Backoff limit per index](#backoff-limit-per-index)
```

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,51 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+  see also [Backoff limit per index](#backoff-limit-per-index)
+- the number of failed indexes in the Job using has exceeded the specified
+  `spec.maxFailedIndexes`; see also [Backoff limit per index](#backoff-limit-per-index)
+- a failed Pod matches a rule in `.spec.podFailurePolicy`
+   with the `FailJob` action; see also [Pod failure policy](#pod-failure-policy)
```
```suggestion
- a failed Pod matches a rule in `.spec.podFailurePolicy`
   with the `FailJob` action; see also [Pod failure policy](#pod-failure-policy) for more details on how the policies can affect evaluation of failures
```

atiratree commented:
/lgtm from sig-apps

shannonxtreme commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,51 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+Job controller awaits for termination of all pods before adding a condition
+indicating that the Job is finished (either `Complete` or `Failed`).
+
+The process of terminating all pods may take a substantial amount
+of time, depending on a Pod's `terminationGracePeriodSeconds` (see
+[Pod termination](#docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)),
+and thus adding the terminal Job condition, even if the fate of the Job is
+already determined.
+
+If you want to know the fate of the Job as soon as determined you can use,
+since v1.31, the `FailureTarget` and `SuccessCriteriaMet` conditions, which
+cover all scenarios in which Job controller triggers the Job termination process
+(see [Terminal Job conditions](#terminal-job-conditions)).
```
```suggestion
Factors like `terminationGracePeriodSeconds` might increase the amount of time
from the moment that the Job controller adds the `FailureTarget` condition or the
`SuccessCriteriaMet` condition to the moment that all of the Job Pods terminate
and the Job controller adds a [terminal condition](#terminal-job-conditions)
(`Failed` or `Complete`).

You can use the `FailureTarget` or the `SuccessCriteriaMet` condition to evaluate
whether the Job has failed or succeeded without having to wait for the controller
to add a terminal condition.
```

shannonxtreme replied:
I think that this is the most important bit of info. You can also add an example or two to the last paragraph for WHY you'd need this, like 

"For example, you can use this information to quickly decide whether to create a replacement Job, or to make scheduling decisions based on node availability."

mimowo replied:
Rephrased, also added the WHY in a note.

shannonxtreme commented on content/en/docs/concepts/workloads/controllers/job.md:
````diff
@@ -1063,6 +1123,13 @@ status:
   terminating: 3 # three Pods are terminating and have not yet reached the Failed phase
 ```
 
+{{< note >}}
+Since v1.31, when you enable the `JobPodReplacementPolicy`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
+(enabled by default), the Job controller awaits for termination of all pods
+before marking a Job as terminal (see [Termination of Job Pods](#termination-of-job-pods)).
+{{< /note >}}
````
Not sure how this relates to the replacement of failed Pods. I would remove this, since it's a bit confusing here

mimowo replied:
I agree, it can be confusing. The only purpose of this was to reference that the logic change is controlled by the feature gate. Now I propose to mention this as a note in one place only.

shannonxtreme commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -1107,6 +1174,13 @@ after the operation: the built-in Job controller and the external controller
 indicated by the field value.
 {{< /warning >}}
 
+{{< note >}}
```
Note not needed in this section, as it's not really related to the section's topic

mimowo replied:
I agree, it can be confusing. The only purpose of this was to reference that the logic change is controlled by the feature gate. Now I propose to mention this as a note in one place only.

shannonxtreme commented on content/en/docs/tasks/job/pod-failure-policy.md:
```diff
@@ -58,6 +58,12 @@ equal `PodFailurePolicy`. Additionally, the `message` field contains a
 more detailed information about the Job termination, such as:
 `Container main for pod default/job-pod-failure-policy-failjob-8ckj8 failed with exit code 42 matching FailJob rule at index 0`.
 
+{{< note >}}
+You will also observe the `FailureTarget` Job condition, with the same `reason`
+and `message` as for the the `Failed` condition (see also
+[Job termination and cleanup](#job-termination-and-cleanup)).
+{{< /note >}}
+
```
Try this:

In the Job status, the following conditions display:

- `Failed` condition: has a `reason` field set to `PodFailurePolicy` and a
   `message` field with more information about the termination, like
   `Container main for pod default/job-pod-failure-policy-failjob-8ckj8 failed with exit code 42 matching FailJob rule at index 0`.
   The Job controller adds this condition after all of the Job's Pods are terminated.
- `FailureTarget` condition: same reason and message as the `Failed` condition. The Job
   controller adds this condition as soon as the Job is considered a failure. For details, see
   [Termination of Job Pods](/docs/concepts/workloads/controllers/job/#termination-of-job-pods).

mimowo replied:
Makes sense, updated along the lines, but first mention `FailureTarget` and then `Failed` to reflect the natural order of them being added.

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -442,8 +442,18 @@ kubectl get -o yaml job job-backoff-limit-per-index-example
+    - message: Job has failed indexes
+      reason: FailedIndexes
+      status: "True"
+      type: FailureTarget # Since v1.31
```
Makes sense, especially as currently the 1.31 version is scattered in a couple of places. I think it makes sense to hint users that there is a change of default behavior, but one place is probably enough,

I have also flipped the order to the conditions to reflect the order of them being added in time.

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -442,8 +442,18 @@ kubectl get -o yaml job job-backoff-limit-per-index-example
+Since v1.31, you will also observe in the status the `FailureTarget`
+Job condition, with the same `reason` and `message` as for the the `Failed`
+condition (see also [Job termination and cleanup](#job-termination-and-cleanup)).
+{{< /note >}}
```
Makes sense, but I yet rephrased it a bit to make it more clear that it is `Failed` condition that takes reason and message from `FailureTarget`, not the other way round. 

Also, the case shown in this example is not describing failure due to pod failure policy, but backoff limit per index.

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,51 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+
+A Job has two possible terminal states, it ends up either succeeded, or failed,
+and these states are reflected by the presence of the Job conditions `Complete`
+or `Failed`, respectively.
```
done

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,51 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+and these states are reflected by the presence of the Job conditions `Complete`
+or `Failed`, respectively.
+
+The failure scenarios encompass:
```
Actually, the lists are complete so I prefer to have a wording which reflects that. How about "The possible reasons for a Job failure / success". Tentatively updated this way.

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,51 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+- a failed Pod matches a rule in `.spec.podFailurePolicy`
+   with the `FailJob` action; see also [Pod failure policy](#pod-failure-policy) for more details on how the policy rules can affect evaluation of failures
+
+The success scenarios encompass:
```
Rephrased tentatively to "The possible reasons for a Job success" to better reflect this the complete list of scenarios.

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,51 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+
+### Termination of Job pods
+
+Prior to v1.31 the Job terminal conditions are added when the Job termination
+process is triggered, and all Pod finalizers are removed, but some pods may
+still remain running/terminating at that point in time.
```
I think it would be useful to keep trace of the change in one place, as this is a substantial logic change, and even though the docs are versioned users may not have a clue something has changed. We did it in the past for numerous features, for example: https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/.

That being said, I make sure the version is mentioned in only one place (before review it was scattered across).

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,51 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+`JobManagedBy` or the `JobPodReplacementPolicy` (enabled by default)
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/), the
+Job controller awaits for termination of all pods before adding a condition
+indicating that the Job is finished (either `Complete` or `Failed`).
```
Rephrased

mimowo commented:
@shannonxtreme thank you for reviewing, great comments aiming to improve comprehension by users!
@atiratree you may want to give another pass as some of the rephrasing are substantial.

Princesso commented:
Hello @AndrewSirenko ,

v1.31 Doc Lead here. This PR has been marked as at risk for Doc Freeze.

The Doc Reviewers have given feedback on it, however, an update is yet to be made on it, based on feedback.

Please note that the Doc freeze deadline is tomorrow, and if the PR is not merged by tomorrow, you will need to file an exception or this enhancement will be removed from this release. Thank you!

sftim commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,66 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+until all pods are terminated.
+
+{{< note >}}
+In the previous releases Job terminal conditions are added when the Job
```
```suggestion
In Kubernetes v1.30 and earlier, Job terminal conditions were added when the Job
```
Use past tense ideally.

sftim commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,66 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+- the criteria specified by the `.spec.successPolicy` are met;
+  see also [Success policy](#success-policy)
+
+Since v1.31 the Job controller delays the addition of the terminal conditions
```
```suggestion
In Kubernetes v1.31 and later the Job controller delays the addition of the terminal conditions
```
?

atiratree commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,66 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+- the criteria specified by the `.spec.successPolicy` are met;
+  see also [Success policy](#success-policy)
+
+Since v1.31 the Job controller delays the addition of the terminal conditions
```
```suggestion
Since v1.31 the Job controller delays the addition of the terminal conditions `Complete` or `Failed`
```

I think it wouldn't hurt to reiterate the conditions.

mimowo replied:
sure, done

atiratree reviewed (commented):
mostly lgtm, but there are still unresolved reviews (e.g. the failure/success lists).

mimowo commented:
> mostly lgtm, but there are still unresolved reviews (e.g. the failure/success lists).

Right, I believe I addressed all comments, but I didn't want to mark some of the comments (especially recent from @shannonxtreme ) as resolved so that there is visibility into the recent changes and discussions.

mimowo commented:
> Right, I believe I addressed all comments, but I didn't want to mark some of the comments (especially recent from @shannonxtreme ) as resolved so that there is visibility into the recent changes and discussions.

It turned out that I missed some comments indeed (or somehow revered the automatically applied changes). I have manually re-applied the comments. Hope I didn't miss anything this time. PTAL.

shannonxtreme commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,67 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+terminal conditions,`Failed` or `Succeeded`, until all pods are terminated.
+
+{{< note >}}
+In Kubernetes v1.30 and earlier, Job terminal conditions were added when the Job
+termination process is triggered, and all Pod finalizers are removed, but some
+pods may still remain running/terminating at that point in time.
+
+The change of the behavior is activated by enablement of the `JobManagedBy` or
+`JobPodReplacementPolicy` (enabled by default)
+[feature gates](/docs/reference/command-line-tools-reference/feature-gates/).
```
```suggestion
In Kubernetes v1.30 and earlier, the Job controller added the `Complete` or the
`Failed` Job terminal conditions as soon as the Job termination process was
triggered and all Pod finalizers were removed. However, some Pods would still
be running or terminating at the moment that the terminal condition was added.

In Kubernetes v1.31 and later, the controller only adds the Job terminal conditions
_after_ all of the Pods are terminated. You can enable this behavior by using the `JobManagedBy` or the `JobPodReplacementPolicy` (enabled by default)
[feature gates](/docs/reference/command-line-tools-reference/feature-gates/).
```

shannonxtreme commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,67 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+In Kubernetes v1.31 and later the Job controller delays the addition of the
+terminal conditions,`Failed` or `Succeeded`, until all pods are terminated.
+
+{{< note >}}
```
I think this doesn't belong in a note, because the previous sentence is also related to the content inside the note.

shannonxtreme replied:
Or bring the preceding sentence about 1.31 and later into the note block

mimowo replied:
Ok, I chose to remove the note entirely

shannonxtreme commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,67 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+  [Success policy](#success-policy).
+
+In Kubernetes v1.31 and later the Job controller delays the addition of the
+terminal conditions,`Failed` or `Succeeded`, until all pods are terminated.
```
```suggestion
terminal conditions,`Failed` or `Complete`, until all of the Job Pods are terminated.
```

sftim reviewed (commented):
LGTM for docs

shannonxtreme commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,67 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+whether the Job has failed or succeeded without having to wait for the controller
+to add a terminal condition.
+
+{{< note >}}
```
No need a note callout; the example flows nicely with the rest of the section

mimowo replied:
Ok, removed.

shannonxtreme commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,67 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+to add a terminal condition.
+
+{{< note >}}
+For example, you can use the `FailureTarget` condition to quickly decide whether
+to create a replacement Job, but it could result in Pods from the failing and
+replacement Jobs running at the same time for a while. Thus, if your cluster
+capacity is limited, you may prefer to wait for the `Failed` condition before
+creating the replacement Job.
```
```suggestion
For example, you might want to decide when to create a replacement Job
that replaces a failed Job. If you replace the failed Job when the `FailureTarget`
condition appears, your replacement Job runs sooner, but could result in Pods
from the failed and the replacement Job running at the same time, using
extra compute resources. 

Alternatively, if your cluster has limited resource capacity, you could choose to
wait until the `Failed` condition appears on the Job, which would delay your
replacement Job but would ensure that you conserve resources by waiting
until all of the failed Pods are removed.
```

mimowo commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -659,6 +674,67 @@ there is no automatic Job restart once the Job status is `type: Failed`.
+   action. For details about how Pod failure policy rules might affect failure
+   evaluation, see [Pod failure policy](#pod-failure-policy).
+
+The possible reasons for a Job success:
```
done

mimowo commented:
@shannonxtreme thank you for reviewing. I believe all comments are addressed. PTAL.

mimowo commented:
@tengqm would you like to approve?

shannonxtreme commented:
Thanks @mimowo !

/lgtm

Princesso commented:
/approve
